pkg/envtest/apiserver.go

```diff
@@ -0,0 +1,443 @@
+func (a *GardenerAPIServer) prepareKubeconfigFile() (string, error) {
+	kubeconfigBytes, err := util.MarshalKubeconfigWithClientCertificate(a.restConfig, nil, nil)
+	if err != nil {
+		return "", err
+	}
+	kubeconfigFile := filepath.Join(a.CertDir, "kubeconfig.yaml")
+
+	return kubeconfigFile, ioutil.WriteFile(kubeconfigFile, kubeconfigBytes, 0600)
+}
+
+// waitUntilHealthy waits for the HealthCheckEndpoint to return 200.
+func (a *GardenerAPIServer) waitUntilHealthy(ctx context.Context) error {
+	// setup secure http client
+	certPool := x509.NewCertPool()
+	certPool.AppendCertsFromPEM(a.caCert.CertificatePEM)
+	httpClient := &http.Client{Transport: &http.Transport{TLSClientConfig: &tls.Config{RootCAs: certPool}}}
+
+	healthCheckURL := a.listenURL
+	healthCheckURL.Path = a.HealthCheckEndpoint
+
+	err := retry.Until(ctx, waitPollInterval, func(context.Context) (bool, error) {
+		res, err := httpClient.Get(healthCheckURL.String())
+		if err == nil {
+			_ = res.Body.Close()
+			if res.StatusCode == http.StatusOK {
+				log.V(1).Info("gardener-apiserver got healthy")
+				return retry.Ok()
+			}
+		}
+		return retry.MinorError(err)
+	})
+	if err != nil {
+		if stopErr := a.Stop(); stopErr != nil {
+			log.Error(stopErr, "failed stopping gardener-apiserver")
+		}
+	}
+	return err
+}
+
+var allGardenerAPIGroupVersions = []schema.GroupVersion{
+	gardencorev1beta1.SchemeGroupVersion,
+	gardencorev1alpha1.SchemeGroupVersion,
+	settingsv1alpha1.SchemeGroupVersion,
+	seedmanagementv1alpha1.SchemeGroupVersion,
+}
+
+// registerGardenerAPIs registers GardenerAPIServer's APIs in the test environment and waits for them to be discoverable.
+func (a *GardenerAPIServer) registerGardenerAPIs(ctx context.Context) error {
+	c, err := client.New(a.restConfig, client.Options{Scheme: kubernetes.GardenScheme})
+	if err != nil {
+		return err
+	}
+
+	// create ExternalName service pointing to localhost
+	service := &corev1.Service{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "gardener-apiserver",
+			Namespace: metav1.NamespaceSystem,
+		},
+		Spec: corev1.ServiceSpec{
+			Type:         corev1.ServiceTypeExternalName,
+			ExternalName: "localhost",
+		},
+	}
+	if err := c.Create(ctx, service); err != nil {
+		return err
+	}
+
+	// create APIServices for all API GroupVersions served by GardenerAPIServer
+	var allAPIServices []*apiregistrationv1.APIService
+	for _, gv := range allGardenerAPIGroupVersions {
+		apiService := a.apiServiceForSchemeGroupVersion(service, gv)
+		allAPIServices = append(allAPIServices, apiService)
+		if err := c.Create(ctx, apiService); err != nil {
+			return err
+		}
+	}
+
+	// wait for all the APIServices to be available
+	if err := retry.Until(ctx, waitPollInterval, func(ctx context.Context) (bool, error) {
+		for _, apiService := range allAPIServices {
+			if err := c.Get(ctx, client.ObjectKeyFromObject(apiService), apiService); err != nil {
+				return retry.MinorError(err)
```

Comment:
Shouldn't this be a SevereError?

Reply:
I think, it's better to give it another try when we receive some error here. If we exit early in that case, test suites might fail sporadically under cpu pressure or whatever, although the situation might recover with a bit more patience :)
Also, a few retries shouldn't hurt when executing locally, I guess, as the default startup timeout is `20s` (`1m` if executed via `hack/test[-cover].sh`).

Reply:
OK, makes sense as it's for unit tests anyways 👍🏻 Thanks!